cabby/model/text/dual_encoder/train.py

```diff
@@ -30,164 +30,195 @@
+
+    # Validation loop.
+    logging.info("Starting evaluation.")
+    true_points_list, pred_points_list = [], []
+    predictions_list, true_vals  = [], []
+    correct = 0
+    total = 0
+    loss_val_total = 0
+
+    for batch in self.valid_loader:
+      text = {key: val.to(self.device) for key, val in batch['text'].items()}
+      cellids = batch['cellid'].float().to(self.device)
+      neighbor_cells = batch['neighbor_cells'].float().to(self.device) 
+      far_cells = batch['far_cells'].float().to(self.device)
+
+      cellids = cellids.float().to(self.device) 
+      neighbor_cells = neighbor_cells.float().to(self.device) 
+      far_cells = far_cells.float().to(self.device)
+      text = {key: val.to(self.device) for key, val in text.items()}
+
+      loss = self.compute_loss(text, cellids, neighbor_cells, far_cells)
+
+      loss_val_total += loss.item()
+
+      text_embedding, cellid_embedding = self.model(text, self.cells_tensor)
+      batch_dim = text_embedding.shape[0]
+      cell_dim = cellid_embedding.shape[0]
+      output_dim  = cellid_embedding.shape[1]
+      cellid_embedding_exp = cellid_embedding.expand(
+        batch_dim, cell_dim, output_dim)
+      text_embedding_exp = text_embedding.unsqueeze(1)
+      output = self.cos(cellid_embedding_exp,text_embedding_exp)
+      output = output.detach().cpu().numpy()
+      predictions = np.argmax(output, axis=1)
+      predictions_list.append(predictions)
+      labels = batch['label'].numpy()
+      true_vals.append(labels)
+      true_points_list.append(batch['point'])
+
+    true_points_list = np.concatenate(true_points_list, axis=0)
+    predictions_list = np.concatenate(predictions_list, axis=0)
+    pred_points_list = util.predictions_to_points(
+      predictions_list, self.label_to_cellid)
+    true_vals = np.concatenate(true_vals, axis=0)
+    average_valid_loss = loss_val_total / len(self.valid_loader)
+
+    return (average_valid_loss, predictions_list, true_vals, 
+    true_points_list, pred_points_list)
+
+  def compute_loss(
+          self,
+          text: Dict, 
+          cellids: torch.tensor, 
+          neighbor_cells: torch.tensor, 
+          far_cells: torch.tensor):
+    
+    # Correct cellid.
+    target = torch.ones(cellids.shape[0]).to(self.device)
+    text_embedding, cellid_embedding = self.model(text, cellids)
+    loss_cellid = criterion(text_embedding, cellid_embedding, target)
+
+    # Neighbor cellid.
+    target_neighbor = -1*torch.ones(cellids.shape[0]).to(self.device)
+    text_embedding_neighbor, cellid_embedding = self.model(text, neighbor_cells)
+    loss_neighbor = criterion(text_embedding_neighbor, 
+    cellid_embedding, target_neighbor)
+
+    # Far cellid.
+    target_far = -1*torch.ones(cellids.shape[0]).to(self.device)
+    text_embedding_far, cellid_embedding = self.model(text, far_cells)
+    loss_far = criterion(text_embedding_far, cellid_embedding, target_far)
+
+    loss = loss_cellid + loss_neighbor + loss_far
```

Comment:
maybe add a positive loss weight? 
_num_negatives = 2
_positive_weight = 0.7
loss = (_positive_weight * loss_cellid) + (1-_positive_weight) * (loss_neighbor + loss_far) / _num_negatives

Reply:
I tried that it did not improve

Reply:
ack! good to know! do you have some place to track these variations you have been trying? Would be good to know what worked, what didnt etc. We usually keep a sheet to track the experiments for specific models e.g. see go/mage-geocoding-data 'experiments - 2020' sheet.